### ISSUE TITLE
correlation with confounds after confound regression

### ISSUE THREAD

**Issue by fliem** [nilearn/nilearn#1064]
I think this one is related to changes introduced in #963.
When regressing out confouds from a signal with standardize=False, correlations between the signal and the confounds still presist (3rd panel), while they are eliminated if standardize=True (2nd panel). 
Is that expected behaviour?

```
import numpy as np
from nilearn.signal import clean
import matplotlib.pyplot as plt

r=[]
r_w_std=[]
r_no_std=[]
for i in range(1000):
    X=np.random.rand(100,1)
    c = X + np.random.rand(100,1)
    X_w_std = clean(X, detrend=False, standardize=True, confounds=c)
    X_no_std = clean(X, detrend=False, standardize=False, confounds=c)                                

    r.append(np.corrcoef(X.T,c.T)[1,0])
    r_w_std.append(np.corrcoef(X_w_std.T,c.T)[1,0])
    r_no_std.append(np.corrcoef(X_no_std.T,c.T)[1,0])

f, (ax1, ax2,ax3) = plt.subplots(3)
ax1.hist(r)
ax1.set_title('original correlation (signal x confound)')
ax2.hist(r_w_std)
ax2.set_title('correlation after confound regression (standardize=True)')
ax3.hist(r_no_std)
ax3.set_title('correlation after confound regression (standardize=False)')
plt.tight_layout()
plt.show()
```

![confounds](https://cloud.githubusercontent.com/assets/5772811/14011445/e92780c4-f19e-11e5-9812-b33630f2df78.png)

**Comment by eickenberg** [comment]
This may have to do with the fact that uncorrelated != orthogonal.
Uncorrelated is basically orthogonal after centering.
"Removing the confounds" means orthogonalizing with respect to them.

If I am not completely wrong, working in the standardized setting makes
these notions equivalent again, because the data are then centered. This
would explain your observation.

**Comment by salma1601** [comment]
> I think this one is related to changes introduced in #963.

@fliem You are right.
I think we should discuss about the use cases and what we expect from setting `standardize=False` and `detrend=False`. Here is my point

- If no confounds to be removed, I expect my raw signal, non centred and with its original variance
- If confounds to be removed, I expect them to be removed correctly, i.e. **centred confounds orthogonal to centred signal**.  I don't expect my signal variance to be manipulated apart from this operation.
The tricky part is about the mean. I personally don't expect my output signal to be centred when setting `detrend=False`, unless a constant confound is included among the confounds.

What do you think @fliem @AlexandreAbraham @GaelVaroquaux ?

**Comment by eickenberg** [comment]
Why do you say "correctly"? Isn't this a question of convention/consensus?
Is there one? I guess there is, and it is probably what you mention, but I
never know these things.

If you want centeredness in the end to depend on whether there was a
constant confound, you cannot do the following: center, do confound
removal, add center back - because that would disrespect the constant
regressor or any other on that doesn't have zero sum.

What you could do it divide the signal into m\mathbb{1} + \bar s, where
\bar s = (s - m), where s is the signal and m is its mean. Then you can
orthogonalize \bar s wrt centered drifts and m\mathbb{1} wrt
\mathbb{1}\mathbb{1}^T C where C are the confounds. If there is any one
that has a constant part, then this projected version will go to 0,
otherwise not. These would be two separate orthogonalizations, so it may be
a little weird ...

Isn't that a bit complicated? Wouldn't it be better to keep the simplest
possible thing, which is orthogonalization, and explain it very clearly
along with how to about obtaining alternative ways of removing drifts,
instead of having the software guess at what the user wants?

**Comment by AlexandreAbraham** [comment]
For me, this is the same problem as #374, ie should we consider a "baseline" for the signal or not.

When using a high-pass filter, the constant part of the signal (very low trend) is removed too. FSL avoid this by taking into account a "baseline" for the signal. Basically, it boils down to remove the constant trend, apply the filtering and add the trend back again.

Here we have the same problem. Should we consider a "baseline" for the signal somehow?

Honestly, I have no idea on what is the best thing to do.

**Comment by eickenberg** [comment]
Just a thought: I guess a baseline constant also goes along with a baseline
variance - it doesn't make sense to have a baseline of 35000 and unit
variance. I'm not saying this is done. It's just a comment.

**Comment by salma1601** [comment]
>  Wouldn't it be better to ... explain it very clearly along with how to about obtaining alternative ways of removing drifts, instead of having the software guess at what the user wants?

Explaining can be a possible solution, but intuition is important for me. I already get caught in this trap with the ADHD dataset: since linear trend and constant confounds are included in the confounds files, I used `detrend=False`, and since I care about variances, I set `standardize=False`. This gave unexpected results...
From my personal point (for what it is, just to explain my point), intuition for me is not the math part nor the software coding part behind the tool, but what people will use the tool for in the most "standard" cases. But I agree that setting `detrend=False` is not that standard :).

**Comment by NicolasGensollen** [comment]
I believe this has been addressed in #2609. The proposed solution was to separate the standardization of the confounds from the one of the signal through a `standardize_confounds` argument in `signal.clean()` defaulting to `True`.

So, updating the code from the first message gives:

```python
import numpy as np
from nilearn.signal import clean
import matplotlib.pyplot as plt

r=[]
r_w_std=[]
r_no_std=[]
for i in range(1000):
    X=np.random.rand(100,1)
    c = X + np.random.rand(100,1)
    X_w_std = clean(X, detrend=False, standardize=True, standardize_confounds=True, confounds=c)
    # Use standardize_confounds=False to get previous behaviour...
    X_no_std = clean(X, detrend=False, standardize=False, standardize_confounds=True, confounds=c)                                

    r.append(np.corrcoef(X.T,c.T)[1,0])
    r_w_std.append(np.corrcoef(X_w_std.T,c.T)[1,0])
    r_no_std.append(np.corrcoef(X_no_std.T,c.T)[1,0])

f, (ax1, ax2,ax3) = plt.subplots(3)
ax1.hist(r)
ax1.set_title('original correlation (signal x confound)')
ax2.hist(r_w_std)
ax2.set_title('correlation after confound regression (standardize=True)')
ax3.hist(r_no_std)
ax3.set_title('correlation after confound regression (standardize=False)')
plt.tight_layout()
plt.show()
```

![image](https://user-images.githubusercontent.com/2639645/103540921-d2976780-4e9a-11eb-8514-65c09033a1a2.png)

**Comment by bthirion** [comment]
LGTM, thx. Closing.
